File: src/brs/Account.java
package brs;

import static brs.Constants.AT_PUBLIC_KEY_BYTES;

import brs.crypto.Crypto;
import brs.crypto.EncryptedData;
import brs.db.SignumKey;
import brs.db.VersionedBatchEntityTable;
import brs.util.Convert;
import java.util.Arrays;
import java.util.logging.Level;
import java.util.logging.Logger;

//TODO: Create JavaDocs and remove this
@SuppressWarnings({ "checkstyle:MissingJavadocTypeCheck", "checkstyle:MissingJavadocMethodCheck" })

public class Account {

    private static final Logger logger = Logger.getLogger(Account.class.getSimpleName());

    public final long id;
    public final SignumKey nxtKey;
    private final int creationHeight;
    private byte[] publicKey;
    private int keyHeight;
    private boolean isAutomatedTransaction;

    protected String name;
    protected String description;

    public static class Balance {
        public final long id;
        public final SignumKey nxtKey;

        protected long balanceNqt;
        protected long unconfirmedBalanceNqt;
        protected long forgedBalanceNqt;

        public Balance(long id) {
            this.id = id;
            this.nxtKey = accountSignumKeyFactory().newKey(this.id);
        }

        public void setForgedBalanceNqt(long forgedBalanceNqt) {
            this.forgedBalanceNqt = forgedBalanceNqt;
        }

        public void setUnconfirmedBalanceNqt(long unconfirmedBalanceNqt) {
            this.unconfirmedBalanceNqt = unconfirmedBalanceNqt;
        }

        public void setBalanceNqt(long balanceNqt) {
            this.balanceNqt = balanceNqt;
        }

        public long getId() {
            return id;
        }

        public long getBalanceNqt() {
            return balanceNqt;
        }

        public long getUnconfirmedBalanceNqt() {
            return unconfirmedBalanceNqt;
        }

        public long getForgedBalanceNqt() {
            return forgedBalanceNqt;
        }

        public void checkBalance() {
            Account.checkBalance(this.id, this.balanceNqt, this.unconfirmedBalanceNqt);
        }

    }

    public void setName(String name) {
        this.name = name;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public void setPublicKey(byte[] publicKey) {
        this.publicKey = publicKey;
    }

    public void setKeyHeight(int keyHeight) {
        this.keyHeight = keyHeight;
    }

    public void setIsAt(boolean isAutomatedTransaction) {
        this.isAutomatedTransaction = isAutomatedTransaction;
    }

    public boolean isAutomatedTransaction() {
        return this.isAutomatedTransaction;
    }

    public enum Event {
        BALANCE, UNCONFIRMED_BALANCE, ASSET_BALANCE, UNCONFIRMED_ASSET_BALANCE,
        LEASE_SCHEDULED, LEASE_STARTED, LEASE_ENDED

    }

    public static class AccountAsset {
        public final long accountId;
        public final long assetId;
        public final SignumKey signumKey;
        private long quantityQnt;
        private long unconfirmedQuantityQnt;
        private boolean isTreasury;

        protected AccountAsset(
                long accountId,
                long assetId,
                long quantityQnt,
                long unconfirmedQuantityQnt,
                SignumKey signumKey) {
            this.accountId = accountId;
            this.assetId = assetId;
            this.quantityQnt = quantityQnt;
            this.unconfirmedQuantityQnt = unconfirmedQuantityQnt;
            this.signumKey = signumKey;
            this.isTreasury = false;
        }

        public AccountAsset(
                SignumKey signumKey,
                long accountId,
                long assetId,
                long quantityQnt,
                long unconfirmedQuantityQnt) {
            this.accountId = accountId;
            this.assetId = assetId;
            this.signumKey = signumKey;
            this.quantityQnt = quantityQnt;
            this.unconfirmedQuantityQnt = unconfirmedQuantityQnt;
            this.isTreasury = false;
        }

        public long getAccountId() {
            return accountId;
        }

        public long getAssetId() {
            return assetId;
        }

        public long getQuantityQnt() {
            return quantityQnt;
        }

        public long getUnconfirmedQuantityQnt() {
            return unconfirmedQuantityQnt;
        }

        public void checkBalance() {
            Account.checkBalance(this.accountId, this.quantityQnt, this.unconfirmedQuantityQnt);
        }

        @Override
        public String toString() {
            return "AccountAsset account_id: "
                    + Convert.toUnsignedLong(accountId)
                    + " asset_id: "
                    + Convert.toUnsignedLong(assetId)
                    + " quantity: "
                    + quantityQnt
                    + " unconfirmedQuantity: "
                    + unconfirmedQuantityQnt;
        }

        public void setQuantityQnt(long quantityQnt) {
            this.quantityQnt = quantityQnt;
        }

        public void setUnconfirmedQuantityQnt(long unconfirmedQuantityQnt) {
            this.unconfirmedQuantityQnt = unconfirmedQuantityQnt;
        }

        public void setTreasury(boolean isTreasury) {
            this.isTreasury = isTreasury;
        }

        public boolean isTreasury() {
            return isTreasury;
        }

    }

    public static class RewardRecipientAssignment {
        public final Long accountId;
        private Long prevRecipientId;
        private Long recipientId;
        private int fromHeight;
        public final SignumKey signumKey;

        public RewardRecipientAssignment(
                Long accountId,
                Long prevRecipientId,
                Long recipientId,
                int fromHeight,
                SignumKey signumKey) {
            this.accountId = accountId;
            this.prevRecipientId = prevRecipientId;
            this.recipientId = recipientId;
            this.fromHeight = fromHeight;
            this.signumKey = signumKey;
        }

        public long getAccountId() {
            return accountId;
        }

        public long getPrevRecipientId() {
            return prevRecipientId;
        }

        public long getRecipientId() {
            return recipientId;
        }

        public int getFromHeight() {
            return fromHeight;
        }

        public void setRecipient(long newRecipientId, int fromHeight) {
            prevRecipientId = recipientId;
            recipientId = newRecipientId;
            this.fromHeight = fromHeight;
        }
    }

    static class DoubleSpendingException extends RuntimeException {

        DoubleSpendingException(String message) {
            super(message);
        }

    }

    private static SignumKey.LongKeyFactory<Account> accountSignumKeyFactory() {
        return Signum.getStores().getAccountStore().getAccountKeyFactory();
    }

    private static SignumKey.LongKeyFactory<Account.Balance> accountBalanceSignumKeyFactory() {
        return Signum.getStores().getAccountStore().getAccountBalanceKeyFactory();
    }

    private static VersionedBatchEntityTable<Account> accountTable() {
        return Signum.getStores().getAccountStore().getAccountTable();
    }

    private static VersionedBatchEntityTable<Account.Balance> accountBalanceTable() {
        return Signum.getStores().getAccountStore().getAccountBalanceTable();
    }

    public static Account getAccount(long id) {
        return id == 0 ? null : accountTable().get(accountSignumKeyFactory().newKey(id));
    }

    public static Account.Balance getAccountBalance(long id) {
        return id == 0 ? null
                : accountBalanceTable()
                        .get(accountBalanceSignumKeyFactory()
                                .newKey(id));
    }

    public static Account.AccountAsset getAccountAssetBalance(long id, long assetId) {
        return Signum.getStores().getAccountStore().getAccountAsset(id, assetId);
    }

    public long getId() {
        return id;
    }

    public static long getId(byte[] publicKey) {
        byte[] publicKeyHash = Crypto.sha256().digest(publicKey);
        return Convert.fullHashToId(publicKeyHash);
    }

    public static Account getOrAddAccount(long id) {
        Account account = getAccount(id);
        if (account == null) {
            account = new Account(id);
            accountTable().insert(account);
        }
        return account;
    }

    public Account(long id) {
        if (id != Crypto.rsDecode(Crypto.rsEncode(id))) {
            logger.log(Level.INFO, "CRITICAL ERROR: Reed-Solomon encoding fails for {0}", id);
        }
        this.id = id;
        this.nxtKey = accountSignumKeyFactory().newKey(this.id);
        this.creationHeight = Signum.getBlockchain().getHeight();
    }

    protected Account(long id, SignumKey signumKey, int creationHeight) {
        if (id != Crypto.rsDecode(Crypto.rsEncode(id))) {
            logger.log(Level.INFO, "CRITICAL ERROR: Reed-Solomon encoding fails for {0}", id);
        }
        this.id = id;
        this.nxtKey = signumKey;
        this.creationHeight = creationHeight;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public byte[] getPublicKey() {
        if (this.keyHeight == -1) {
            return null;
        }
        return publicKey;
    }

    public int getCreationHeight() {
        return creationHeight;
    }

    public int getKeyHeight() {
        return keyHeight;
    }

    public long getUnconfirmedBalanceNqt() {
        Balance balance = Account.getAccountBalance(id);
        return balance == null ? 0L : balance.getUnconfirmedBalanceNqt();
    }

    public long getBalanceNqt() {
        Balance balance = Account.getAccountBalance(id);
        return balance == null ? 0L : balance.getBalanceNqt();
    }

    public long getForgedBalanceNqt() {
        Balance balance = Account.getAccountBalance(id);
        return balance == null ? 0L : balance.getForgedBalanceNqt();
    }

    public EncryptedData encryptTo(byte[] data, String senderSecretPhrase) {
        if (getPublicKey() == null) {
            throw new IllegalArgumentException("Recipient account doesn't have a public key set");
        }
        return EncryptedData.encrypt(data, Crypto.getPrivateKey(senderSecretPhrase), publicKey);
    }

    public static EncryptedData encryptTo(
            byte[] data,
            String senderSecretPhrase,
            byte[] publicKey) {
        if (publicKey == null) {
            throw new IllegalArgumentException("public key required");
        }
        return EncryptedData.encrypt(data, Crypto.getPrivateKey(senderSecretPhrase), publicKey);
    }

    public byte[] decryptFrom(EncryptedData encryptedData, String recipientSecretPhrase) {
        if (getPublicKey() == null) {
            throw new IllegalArgumentException("Sender account doesn't have a public key set");
        }
        return encryptedData.decrypt(Crypto.getPrivateKey(recipientSecretPhrase), publicKey);
    }

    // returns true iff:
    // this.publicKey is set to null (in which case this.publicKey also gets set to
    // key)
    // or
    // this.publicKey is already set to an array equal to key
    public boolean setOrVerify(byte[] key, int height) {
        return Signum.getStores().getAccountStore().setOrVerify(this, key, height);
    }

    public void apply(byte[] key, int height) {
        if (!setOrVerify(key, this.creationHeight)) {
            throw new IllegalStateException("Public key mismatch");
        }
        if (this.publicKey == null) {
            throw new IllegalStateException(
                    "Public key has not been set for account " + Convert.toUnsignedLong(id)
                            + " at height " + height + ", key height is " + keyHeight);
        }
        if (this.keyHeight == -1 || this.keyHeight > height) {
            this.keyHeight = height;
            accountTable().insert(this);
        }
    }

    public static boolean checkIsAutomatedTransaction(Account account) {
        return Arrays.equals(account.getPublicKey(), AT_PUBLIC_KEY_BYTES);
    }

    private static void checkBalance(long accountId, long confirmed, long unconfirmed) {
        if (confirmed < 0) {
            throw new DoubleSpendingException("Negative balance or quantity ("
                    + confirmed
                    + ") for account "
                    + Convert.toUnsignedLong(accountId));
        }
        if (unconfirmed < 0) {
            throw new DoubleSpendingException("Negative unconfirmed balance or quantity ("
                    + unconfirmed
                    + ") for account "
                    + Convert.toUnsignedLong(accountId));
        }
        if (unconfirmed > confirmed) {
            throw new DoubleSpendingException("Unconfirmed ("
                    + unconfirmed
                    + ") exceeds confirmed ("
                    + confirmed
                    + ") balance or quantity for account "
                    + Convert.toUnsignedLong(accountId));
        }
    }

}

File: src/main/kotlin/brs/entity/Account.kt
package brs.entity

import brs.db.BurstKey
import brs.db.VersionedBatchEntityTable
import brs.util.convert.fullHashToId
import brs.util.convert.toUnsignedString
import brs.util.crypto.Crypto
import brs.util.crypto.rsVerify
import java.util.logging.Level
import java.util.logging.Logger

open class Account {
    val id: Long
    val nxtKey: BurstKey
    val creationHeight: Int
    private var publicKeyInternal: ByteArray? = null
    var publicKey: ByteArray?
        get() = if (this.keyHeight == -1) {
            null
        } else publicKeyInternal
        set(v) {
            publicKeyInternal = v
        }
    var keyHeight: Int = 0
    var balancePlanck: Long = 0
    var unconfirmedBalancePlanck: Long = 0
    var forgedBalancePlanck: Long = 0

    var name: String? = null
    var description: String? = null

    enum class Event {
        BALANCE, UNCONFIRMED_BALANCE, ASSET_BALANCE, UNCONFIRMED_ASSET_BALANCE
    }

    open class AccountAsset {
        val accountId: Long
        val assetId: Long
        val burstKey: BurstKey
        var quantity: Long = 0
        var unconfirmedQuantity: Long = 0

        protected constructor(
            accountId: Long,
            assetId: Long,
            quantity: Long,
            unconfirmedQuantity: Long,
            burstKey: BurstKey
        ) {
            this.accountId = accountId
            this.assetId = assetId
            this.quantity = quantity
            this.unconfirmedQuantity = unconfirmedQuantity
            this.burstKey = burstKey
        }

        constructor(burstKey: BurstKey, accountId: Long, assetId: Long, quantity: Long, unconfirmedQuantity: Long) {
            this.accountId = accountId
            this.assetId = assetId
            this.burstKey = burstKey
            this.quantity = quantity
            this.unconfirmedQuantity = unconfirmedQuantity
        }

        fun checkBalance() {
            checkBalance(this.accountId, this.quantity, this.unconfirmedQuantity)
        }

        override fun toString(): String {
            return ("AccountAsset account_id: "
                    + accountId.toUnsignedString()
                    + " asset_id: "
                    + assetId.toUnsignedString()
                    + " quantity: "
                    + quantity
                    + " unconfirmedQuantity: "
                    + unconfirmedQuantity)
        }
    }

    open class RewardRecipientAssignment(
        val accountId: Long,
        var prevRecipientId: Long,
        var recipientId: Long,
        fromHeight: Int,
        val burstKey: BurstKey
    ) {
        var fromHeight: Int = 0
            private set

        init {
            this.fromHeight = fromHeight
        }

        fun setRecipient(newRecipientId: Long, fromHeight: Int) {
            prevRecipientId = recipientId
            recipientId = newRecipientId
            this.fromHeight = fromHeight
        }
    }

    internal class DoubleSpendingException(message: String) : RuntimeException(message)

    constructor(dp: DependencyProvider, id: Long) {
        if (!id.rsVerify()) {
            logger.log(Level.INFO, "CRITICAL ERROR: Reed-Solomon encoding fails for {0}", id)
        }
        this.id = id
        this.nxtKey = accountBurstKeyFactory(dp).newKey(this.id)
        this.creationHeight = dp.blockchainService.height
    }

    protected constructor(id: Long, burstKey: BurstKey, creationHeight: Int) {
        if (!id.rsVerify()) {
            logger.log(Level.INFO, "CRITICAL ERROR: Reed-Solomon encoding fails for {0}", id)
        }
        this.id = id
        this.nxtKey = burstKey
        this.creationHeight = creationHeight
    }

    fun encryptTo(data: ByteArray, senderSecretPhrase: String): EncryptedData {
        requireNotNull(publicKey) { "Recipient account doesn't have a public key set" }
        return EncryptedData.encrypt(data, Crypto.getPrivateKey(senderSecretPhrase), publicKey!!)
    }


    fun decryptFrom(encryptedData: EncryptedData, recipientSecretPhrase: String): ByteArray {
        requireNotNull(publicKey) { "Sender account doesn't have a public key set" }
        return encryptedData.decrypt(Crypto.getPrivateKey(recipientSecretPhrase), publicKey!!)
    }

    // returns true if:
    // this.publicKey is set to null (in which case this.publicKey also gets set to key)
    // or
    // this.publicKey is already set to an array equal to key
    fun setOrVerify(dp: DependencyProvider, key: ByteArray, height: Int): Boolean { // TODO should we deprecate?
        return dp.accountStore.setOrVerify(this, key, height)
    }

    fun apply(dp: DependencyProvider, key: ByteArray, height: Int) {
        check(setOrVerify(dp, key, this.creationHeight)) { "Public key mismatch" }
        checkNotNull(this.publicKeyInternal) {
            ("Public key has not been set for account " + id.toUnsignedString()
                    + " at height " + height + ", key height is " + keyHeight)
        }
        if (this.keyHeight == -1 || this.keyHeight > height) {
            this.keyHeight = height
            accountTable(dp).insert(this)
        }
    }

    fun checkBalance() {
        checkBalance(this.id, this.balancePlanck, this.unconfirmedBalancePlanck)
    }

    companion object {

        private val logger = Logger.getLogger(Account::class.java.simpleName)

        // TODO refactor methods that take dp
        private fun accountBurstKeyFactory(dp: DependencyProvider): BurstKey.LongKeyFactory<Account> {
            return dp.accountStore.accountKeyFactory
        }

        private fun accountTable(dp: DependencyProvider): VersionedBatchEntityTable<Account> {
            return dp.accountStore.accountTable
        }

        @Deprecated("Use dp.accountService.getAccount()")
        fun getAccount(dp: DependencyProvider, id: Long): Account? {
            return if (id == 0L) null else accountTable(dp)[accountBurstKeyFactory(
                dp
            ).newKey(id)]
        }

        @Deprecated("Just use Crypto/Convert class instead")
        fun getId(publicKey: ByteArray): Long {
            val publicKeyHash = Crypto.sha256().digest(publicKey)
            return publicKeyHash.fullHashToId()
        }

        fun getOrAddAccount(dp: DependencyProvider, id: Long): Account {
            var account = getAccount(dp, id)
            if (account == null) {
                account = Account(dp, id)
                accountTable(dp).insert(account)
            }
            return account
        }

        fun encryptTo(data: ByteArray, senderSecretPhrase: String, publicKey: ByteArray?): EncryptedData {
            requireNotNull(publicKey) { "public key required" }
            return EncryptedData.encrypt(data, Crypto.getPrivateKey(senderSecretPhrase), publicKey)
        }

        private fun checkBalance(accountId: Long, confirmed: Long, unconfirmed: Long) {
            if (confirmed < 0) {
                throw DoubleSpendingException(
                    "Negative balance or quantity ("
                            + confirmed
                            + ") for account "
                            + accountId.toUnsignedString()
                )
            }
            if (unconfirmed < 0) {
                throw DoubleSpendingException(
                    "Negative unconfirmed balance or quantity ("
                            + unconfirmed
                            + ") for account "
                            + accountId.toUnsignedString()
                )
            }
            if (unconfirmed > confirmed) {
                throw DoubleSpendingException(
                    "Unconfirmed ("
                            + unconfirmed
                            + ") exceeds confirmed ("
                            + confirmed
                            + ") balance or quantity for account "
                            + accountId.toUnsignedString()
                )
            }
        }
    }
}
